libbpf-tools/syscount.c

```diff
@@ -0,0 +1,455 @@
+	if (err) {
+		if (env.verbose) {
+			warn("can't find comm for pid %d: %s",
+			     pid, strerror(-err));
+		}
+		snprintf(buf, size, "[pid %d]", pid);
+	}
+	return buf;
+}
+
+static const char *agg_col(__u32 key, char *buf, size_t size, int comm_map_fd)
+{
+	char comm_buf[COMM_NAME_SIZE];
+
+	if (env.process) {
+		snprintf(buf, size, "%-6u %-15s",
+			 key, comm(comm_map_fd, key,
+				   comm_buf, sizeof(comm_buf)));
+	} else {
+		syscall_name(key, buf, size);
+	}
+	return buf;
+}
+
+static const char *agg_colname(void)
+{
+	return (env.process) ? "PID    COMM" : "SYSCALL";
+}
+
+static const char *time_colname(void)
+{
+	return (env.milliseconds) ? "TIME (ms)" : "TIME (us)";
+}
+
+static void print_latency_header(void)
+{
+	printf("%-22s %8s %16s\n", agg_colname(), "COUNT", time_colname());
+}
+
+static void print_count_header(void)
+{
+	printf("%-22s %8s\n", agg_colname(), "COUNT");
+}
+
+static void print_latency(struct data_ext_t *vals, size_t count, int comm_map_fd)
+{
+	double div = env.milliseconds ? 1000000.0 : 1000.0;
+	char buf[2 * COMM_NAME_SIZE];
+
+	print_latency_header();
+	for (int i = 0; i < count && i < env.top; i++)
+		printf("%-22s %8llu %16.3lf\n",
+		       agg_col(vals[i].key, buf, sizeof(buf), comm_map_fd),
+		       vals[i].count, vals[i].total_ns / div);
+	printf("\n");
+}
+
+static void print_count(struct data_ext_t *vals, size_t count, int comm_map_fd)
+{
+	char buf[2 * COMM_NAME_SIZE];
+
+	print_count_header();
+	for (int i = 0; i < count && i < env.top; i++)
+		printf("%-22s %8llu\n",
+		       agg_col(vals[i].key, buf, sizeof(buf), comm_map_fd),
+		       vals[i].count);
+	printf("\n");
+}
+
+static void print_timestamp()
+{
+	time_t now = time(NULL);
+	struct tm tm;
+
+	if (localtime_r(&now, &tm))
+		printf("[%02d:%02d:%02d]\n", tm.tm_hour, tm.tm_min, tm.tm_sec);
+	else
+		warn("localtime_r: %s", strerror(errno));
+}
+
+static bool read_vals(int fd, struct data_ext_t *vals, __u32 *count)
+{
+	__u32 keys[MAX_ENTRIES] = { -1 };
+	struct data_t val;
+	__u32 key = -1;
+	__u32 next_key;
+	int i = 0;
+	int err;
+
+	if (!vals || !count || !*count)
+		return true;
+
+	for (key = -1; i < *count; ) {
+		err = bpf_map_get_next_key(fd, &key, &next_key);
+		if (err && errno != ENOENT) {
+			warn("failed to get next key: %s\n", strerror(errno));
+			return false;
+		} else if (err) {
+			break;
+		}
+		key = keys[i++] = next_key;
+	}
+
+	for (int j = 0; j < i; j++) {
+		err = bpf_map_lookup_elem(fd, &keys[j], &val);
+		if (err && errno != ENOENT) {
+			warn("failed to lookup element: %s\n", strerror(errno));
+			return false;
+		}
+		vals[j].count = val.count;
+		vals[j].total_ns = val.total_ns;
+		vals[j].key = keys[j];
+	}
+
+	/* There is a race here: system calls which are represented by keys
+	 * above and happened between lookup and delete will be ignored.  This
+	 * will be fixed in future by using bpf_map_lookup_and_delete_batch,
+	 * but this function is too fresh to use it in bcc. */
```

Comment:
this is not bcc, really :) you can use it, you'll be limiting yourself to very recent kernels, though

Reply:
Ok, what if I do this based on `uname(2)` information (or there is a better way to get linux version)? I am curious if I can actually do race-less code using batch lookup and delete (or I need a `bpf_map_lookup_and_delete_elem` for hashes)